examples/step-49/step-49.cc

```diff
@@ -99,15 +94,53 @@ void print_mesh_info(const Triangulation<dim> &triangulation,
 // discussed in the introduction how to generate it. This follows the same
 // pattern as used in step-5 to load a mesh, although there it was written in
 // a different file format (UCD instead of MSH).
+//
+// We will be utilizing the SphericalManifold function.  We will attach the
+// boundary IDs created in Gmsh to these manifolds and use them for refinement.
+// Unfortunately in gmsh, it's difficult to create nice circles on the interior
+// holes, so we utilize SphericalManifold to find the boundary IDs and round out
+// the edges on the circle.  The refinement will naturally fall onto
+// refine_global().
 void grid_1()
 {
+  const Point<2> Top_right_hole_origin(0.42, 2.0);
+  const Point<2> Bottom_left_hole_origin(-2.1, -1.54);
+
+  const SphericalManifold<2> Top_right_manifold(Top_right_hole_origin);
+  const SphericalManifold<2> Bottom_left_manifold(Bottom_left_hole_origin);
+
   Triangulation<2> triangulation;
 
   GridIn<2> gridin;
   gridin.attach_triangulation(triangulation);
   std::ifstream f("example.msh");
   gridin.read_msh(f);
 
+  // Here is where we get the boundary IDs made in gmsh, which are in the first
+  // coordinate position, and assign them to manifold ids.  With our example, we
+  // have boundary ID 1 on the top right hole and 2 and 3 for the bottom left
+  // hole. We assign both of these boundary IDs 2 because together they make a
+  // circle to match the manifold we assign it later.
+  triangulation.set_all_manifold_ids_on_boundary(1, 1); // top right hole
+  triangulation.set_all_manifold_ids_on_boundary(2,
+                                                 2); // top of bottom left hole
+  triangulation.set_all_manifold_ids_on_boundary(
+    3, 2); // bottom of bottom left hole
+
+  for (const auto &cell : triangulation.cell_iterators())
+    {
+      for (unsigned face = 0; face < GeometryInfo<2>::faces_per_cell; ++face)
+        {
+          if (cell->face(face)->boundary_id() > 0)
```

Comment:
This seems to be incorrect for the id 3 or not (it will use manifold id 3 instead of 2)? Did you check that the output is correct?
I am not sure that what we are doing here is the best way to go about it. The geometry might have holes on the outer part of these cells. I will explain when we chat.